### PR TITLE
drpolicy: validate CIDRs format

### PR DESCRIPTION
**Pull request by rakeshgm**
Validating the CIDRs in DR Policy  and added tests

Signed-off-by: rakeshgm <rakeshgm014@gmail.com>

### CODE REVIEW

File: controllers/drpolicy_controller.go [hunk 1]
```diff
@@ -243,6 +250,36 @@ func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
...
+		cluster := &drpolicy.Spec.DRClusterSet[i]
+		if reason, err := CIDRsFomatValidate(ctx, cluster.CIDRs, log); err != nil {
+			return reason, err
+		}
+	}
+
+	return "", nil
+}
+
+func CIDRsFomatValidate(ctx context.Context, cidrs []string, log logr.Logger) (string, error) {
```
Comment (Madhu-1) [inline controllers/drpolicy_controller.go:265]: no need to have this extra exported function move the logic to `validateCIDRsFormat` function?

File: controllers/drpolicy_controller.go [hunk 2]
```diff
@@ -243,6 +250,36 @@ func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
...
+func validateCIDRsFormat(ctx context.Context, drpolicy *ramen.DRPolicy, log logr.Logger) (string, error) {
+	// validate the CIDRs format
+	for i := range drpolicy.Spec.DRClusterSet {
+		cluster := &drpolicy.Spec.DRClusterSet[i]
+		if reason, err := CIDRsFomatValidate(ctx, cluster.CIDRs, log); err != nil {
+			return reason, err
+		}
+	}
+
+	return "", nil
```
Comment (Madhu-1) [inline controllers/drpolicy_controller.go:262]: just return the error message, reason seems to be redundant?

File: controllers/drpolicy_controller.go [hunk 3]
```diff
@@ -243,6 +250,37 @@ func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
 	return "", nil
 }
 
+func validateCIDRsFormat(ctx context.Context, drpolicy *ramen.DRPolicy, log logr.Logger) (string, error) {
```
Comment (Madhu-1) [inline controllers/drpolicy_controller.go:253]: can this be `validateCIDR` and no need to pass the ctx as input as its not used anywhere?

File: controllers/drpolicy_controller.go (hunk 1 above)
Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: I initially had written with one function. then I checked the `validateS3Profiles` function [here](https://github.com/RamenDR/ramen/blob/main/controllers/drpolicy_controller.go#L217-L230). It was recently split to make it look clean I believe. I can still make the change. What do you think?

File: controllers/drpolicy_controller.go (hunk 2 above)
Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: removed

File: controllers/drpolicy_controller.go (hunk 3 above)
Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: i have removed `ctx`

File: controllers/drpolicy_controller.go [hunk 4]
```diff
@@ -243,6 +250,36 @@ func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
...
+}
+
+func CIDRsFomatValidate(cidrs []string, log logr.Logger) error {
+	invalidCidrs := []string{}
+
+	for i := range cidrs {
+		if _, _, err := net.ParseCIDR(cidrs[i]); err != nil {
+			invalidCidrs = append(invalidCidrs, cidrs[i])
+
+			log.Error(err, "validation error")
```
Comment (Madhu-1) [inline controllers/drpolicy_controller.go:272]: logging is not required here? the caller can take care of logging?

File: controllers/drpolicy_controller.go [hunk 5]
```diff
@@ -243,6 +250,36 @@ func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
...
+		cluster := &drpolicy.Spec.DRClusterSet[i]
+		if err := CIDRsFomatValidate(cluster.CIDRs, log); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func CIDRsFomatValidate(cidrs []string, log logr.Logger) error {
```
Comment (Madhu-1) [inline controllers/drpolicy_controller.go:265]: no need to export this function, can the function be better named? `validateCIDRsFormat` and `CIDRsFomatValidate` functions names seems same.

File: controllers/drpolicy_controller.go (hunk 1 above)
Comment (Madhu-1) [reply controllers/drpolicy_controller.go]: should be fine as the pattern is already available.

File: controllers/drpolicy_controller.go (hunk 5 above)
Comment (Madhu-1) [reply controllers/drpolicy_controller.go:265]: you can add unit tests?

File: controllers/drpolicy_controller.go (hunk 4 above)
Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: logging is removed

File: controllers/drpolicy_controller.go (hunk 5 above)
Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: yes, the functions seem similar. :). I will try to add the unit tests too.

File: controllers/drpolicy_controller.go [hunk 6]
```diff
@@ -243,6 +250,34 @@ func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
...
+	}
+
+	return nil
+}
+
+func CIDRsFomatValidate(cidrs []string) error {
+	invalidCidrs := []string{}
+
+	for i := range cidrs {
+		if _, _, err := net.ParseCIDR(cidrs[i]); err != nil {
```
Comment (Madhu-1) [inline controllers/drpolicy_controller.go:269]: the actual error message should be returned or else difficult to identify what is wrong?

File: controllers/drpolicy_controller.go (hunk 5 above)
Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: functions renamed

File: controllers/drpolicy_controller.go (hunk 6 above)
Comment (ShyamsundarR) [reply controllers/drpolicy_controller.go:269]: To add:
We log errors from external functions, and pass along errors from internal functions upwards. This helps log where we faced an error from an external routine. e.g [here](https://github.com/RamenDR/ramen/blob/ff29bf0d3dbbcd116dc683c3e91d7152e22e8d3b/controllers/volumereplicationgroup_controller.go#L706-L711)

CIDR validation is pretty basic, IOW it is not invoking a request/response like routine, and hence we can short circuit logging in one place, as below where we return the error.

So to address this, please add a log message above the error return (line 275), that logs all invalid CIDRs found, and return a list of invalid CIDRs with a message in the error, say "failed to parse CIDRs %s". 

Now this error message does not yet tell us which drCluster configuration is invalid, and this error message would be present in the status of the DRPolicy reconciler. To fix, in order to make it easier for the user to see the actual error, in the `ValidateCIDRsFormat` function further add required message on which cluster set the invalid CIDRs were found and return that as the error.

Comment (rakeshgm) [reply controllers/drpolicy_controller.go]: Added logging err from `net.ParseCIDR` and also added cluster name in the return error message.

**Review by ShyamsundarR — APPROVED** [review]

**Pull request merged**